### PR TITLE
Verbose output

### LINKED ISSUE #15: Verbose output
Add a verbose flag that produces additional output

- Use the Python logging facility
- Log the name of the tmpdir once
- Don't log the name of the tmpdir every time
- Log file sizes: input, intermediate, output
- Log input split info
- Log keyspace info, this could be useful for later validation of keyspace in P5 unit tests

### PR DESCRIPTION
**Pull request by awdeorio**
Add `-v/--verbose` flag.

Closes #15

To validate, run with and without the verbose flag.  Please give me feedback on the output.

~~Wait to review until #33 is merged.~~

### CODE REVIEW

File: madoop/mapreduce.py [hunk 1]
```diff
@@ -36,12 +40,13 @@ def mapreduce(input_dir, output_dir, map_exe, reduce_exe):
...
     with tempfile.TemporaryDirectory(prefix="madoop-") as tmpdir:
         tmpdir = pathlib.Path(tmpdir)
+        LOGGER.debug("tmpdir=%s", tmpdir)
 
         # Create stage input and output directory
         map_input_dir = tmpdir/'input'
         map_output_dir = tmpdir/'mapper-output'
-        group_output_dir = tmpdir/'grouper-output'
-        reduce_output_dir = tmpdir/'reducer-output'
+        group_output_dir = tmpdir/'reducer-input'
```
Comment (awdeorio) [inline madoop/mapreduce.py:48]: Nitpick on the naming convention here

**Comment by awdeorio** [comment]
This is now ready to go! @jaredzh

**Review by jaredzh — APPROVED** [review]
LGTM. The logging messages are very informative and easy to understand. Only nitpicks might be to change "splits" into "partitions" for the map stage so it is consistent with the language in the group stage. Also, grouping the "partition message" and "unique keys" messages separately instead of interleaving them for the group stage would increase readability? And also another suggestion is that we can also print the output file sizes?

**Comment by awdeorio** [comment]
Great ideas @jaredzh!  Take a look at the new version.  In particular, the grouping stage debug output:

```
INFO: Starting group stage
DEBUG: mapper-output/part-00000 unique_keys=3
DEBUG: mapper-output/part-00001 unique_keys=3
DEBUG: mapper-output all_unique_keys=5
DEBUG: partition mapper-output/part-00000 >> reducer-input/{part-00000,part-00001,part-00002,part-00003}
DEBUG: partition mapper-output/part-00001 >> reducer-input/{part-00000,part-00001,part-00002,part-00003}
DEBUG: empty partition: rm reducer-input/part-00000
DEBUG: sort reducer-input/part-00001
DEBUG: sort reducer-input/part-00002
DEBUG: sort reducer-input/part-00003
DEBUG: reducer-input/part-00001 unique_keys=2
DEBUG: reducer-input/part-00002 unique_keys=1
DEBUG: reducer-input/part-00003 unique_keys=2
DEBUG: reducer-input all_unique_keys=5
```

**Comment by jaredzh** [comment]
LGTM. I made some changes that fixed a bug where the number of map executions outputted was off by 1 and used `>>`  for the partition part for the map stage so that's consistent with the group stage. Go ahead and merge if it's good. @awdeorio

File: madoop/mapreduce.py [hunk 2]
```diff
@@ -202,7 +202,7 @@ def map_stage(exe, input_dir, output_dir):
                     f"Command returned non-zero: "
                     f"{exe} < {input_path} > {output_path}"
                 ) from err
-    LOGGER.info("Finished map executions: %s", i)
+    LOGGER.info("Finished map executions: %s", i+1)
```
Comment (awdeorio) [inline madoop/mapreduce.py]: ohhh good catch.  I nitpicked the solution ac22df3e124124f44405eb5b434be6052e28d91b

File: madoop/mapreduce.py [hunk 3]
```diff
@@ -139,7 +139,7 @@ def prepare_input_files(input_dir, output_dir):
             assert all(i.parent == outparent for i in outpaths)
             outnames = [i.name for i in outpaths]
             logging.debug(
-                "partition %s > %s/{%s}",
+                "partition %s >> %s/{%s}",
```
Comment (awdeorio) [inline madoop/mapreduce.py:142]: Yes, excellent.  Love consistency.

**Pull request merged**